// crocks/Const/index.js
/** @license ISC License (c) copyright 2016 original and current authors */
/** @author Ian Hofmann-Hicks (evil) */

var VERSION = 2

var _equals = require('../core/equals')
var _implements = require('../core/implements')
var _inspect = require('../core/inspect')
var type = require('../core/types').type('Const')
var _type = require('../core/types').typeFn(type(), VERSION)
var fl = require('../core/flNames')

var isFunction = require('../core/isFunction')
var isSameType = require('../core/isSameType')

function Const(x) {
  var obj;

  if(!arguments.length) {
    throw new TypeError('Const: Must wrap something')
  }

  var equals =
    function (m) { return isSameType(Const, m)
      && _equals(x, m.valueOf()); }

  var inspect =
    function () { return ("Const" + (_inspect(x))); }

  var valueOf =
    function () { return x; }

  function concat(method) {
    return function(m) {
      if(!isSameType(Const, m)) {
        throw new TypeError(("Const." + method + ": Const required"))
      }

      return Const(x)
    }
  }

  function map(method) {
    return function(fn) {
      if(!isFunction(fn)) {
        throw new TypeError(("Const." + method + ": Function required"))
      }

      return Const(x)
    }
  }

  function ap(m) {
    if(!isSameType(Const, m)) {
      throw new TypeError('Const.ap: Const required')
    }

    return Const(x)
  }

  function chain(method) {
    return function(fn) {
      if(!isFunction(fn)) {
        throw new TypeError(("Const." + method + ": Function required"))
      }

      return Const(x)
    }
  }

  return ( obj = {
    inspect: inspect, toString: inspect, valueOf: valueOf,
    type: type, equals: equals, ap: ap,
    concat: concat('concat'),
    map: map('map'),
    chain: chain('chain')
  }, obj[fl.equals] = equals, obj[fl.concat] = concat(fl.concat), obj[fl.map] = map(fl.map), obj[fl.chain] = chain(fl.chain), obj['@@type'] = _type, obj.constructor = Const, obj )
}

Const.type = type
Const['@@type'] = _type

Const['@@implements'] = _implements(
  [ 'ap', 'chain', 'concat', 'equals', 'map' ]
)

module.exports = Const
